### PR TITLE
EN-663 Sidebar component

### PR DESCRIPTION
**Pull request by antictor**
A generic sidebar that sits static on the left of the page and expands when the user moves over the top nav section of it.

The notification section is optional and is responsible for revealing a popover, though the application using it is responsible for the content of that popover as this will differ between applications with different notification data.

The user avatar is also optional but will navigate the user to the provided url.

Jul-08-2019 10-54-37

### CODE REVIEW

File: packages/css-framework/src/fragments/_sidebar.scss [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+  text-align: center;
+  display: inline-block;
+  width: $bar-width - $border;
+}
+
+.rn-iconfill {
+  fill: color(neutral, 200);
+}
+
+
```
Comment (m7kvqbe1) [inline packages/css-framework/src/fragments/_sidebar.scss:48]: Do we need the extra line breaks?

File: packages/docs-site/src/library/pages/components/sidebar.md [hunk 2]
```diff
@@ -0,0 +1,281 @@
...
+
+If the application wishes to display the users name, in the form of their initials then an object must be provided to the `user` property, this must include an `initials` property plus a url property expected by the `LinkComponent`, e.g. 'to'.
+
+The Notification section is responsible for displaying a notification icon if notifications are supported by the application, indicating if there are any unread notifications  and toggling the display of a notification popover containing content passed into it. The host application is responsible for generating the markup and styles used within the notification popover.
+
+### Sidebar without notifications
+A simple sidebar with no notification section, simply showing the current user and using the default Link component so a regular anchor tag is used. 
+
+<CodeHighlighter source={`import { Link } from 'react-router-dom'\n
+const user = { initials: 'ZT', href: '/userprofile' }\n
```
Comment (m7kvqbe1) [inline packages/docs-site/src/library/pages/components/sidebar.md]: Maybe this is pedantic but should you be using ZT here?

File: packages/react-component-library/src/fragments/NotificationSidePanel/index.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+const BAR_WIDTH = 60
+const POPOVER_WIDTH = 335
+
+const NotificationSidePanel: React.FC<NotificationSidePanelProps> = ({
+  children,
+  className = '',
+  unreadNotification,
+}) => {
+  const hasNotifications = children !== null
+  const [showNotifications, setShowNotifications] = useState(false)
```
Comment (m7kvqbe1) [inline packages/react-component-library/src/fragments/NotificationSidePanel/index.tsx:20]: Does it make sense to bundle a lot of this into a custom hook?

File: packages/react-component-library/src/fragments/Sidebar/index.tsx [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+  navItems,
+  NotificationsPopoverContent,
+  unreadNotification = false,
+  user,
+}) => {
+  const [open, setOpen] = useState()
+
+  const renderUserLink = ({ initials, ...rest }: UserLinkType): JSX.Element => (
+    <LinkComponent className="rn-sidebar__nav-link" {...rest}>
+      <Avatar
```
Comment (m7kvqbe1) [inline packages/react-component-library/src/fragments/Sidebar/index.tsx:29]: Is `Avatar` independently documented and can it be used elsewhere or is it tightly coupled to the Sidebar at the moment?

File: packages/react-component-library/src/icons/index.ts [hunk 5]
```diff
@@ -1,5 +1,9 @@
+import Bell from './Bell'
+import Graph from './Graph'
+import House from './House'
 import Search from './Search'
+import Tools from './Tools'
 import TriangleUp from './TriangleUp'
 import TriangleDown from './TriangleDown'
 
-export { Search, TriangleDown, TriangleUp }
+export { Bell, Graph, House, Search, Tools, TriangleDown, TriangleUp }
```
Comment (m7kvqbe1) [inline packages/react-component-library/src/icons/index.ts:9]: Are you lifting these from prototype mocks? Are the designers creating these themselves or are we using an off the shelf icon library?

**Review by m7kvqbe1 — APPROVED** [review]
Looks good - few queries but nothing blocking.

File: packages/css-framework/src/fragments/_sidebar.scss (hunk 1 above)
Comment (antictor) [reply packages/css-framework/src/fragments/_sidebar.scss:48]: I put them in just to help organise the file to group styles together

File: packages/react-component-library/src/icons/index.ts (hunk 5 above)
Comment (antictor) [reply packages/react-component-library/src/icons/index.ts:9]: The icon library question still seems to be up in the air

File: packages/react-component-library/src/fragments/Sidebar/index.tsx (hunk 4 above)
Comment (antictor) [reply packages/react-component-library/src/fragments/Sidebar/index.tsx:29]: Makes sense to document it on it's own as it's a component. I'll do that in a follow up

File: packages/react-component-library/src/fragments/NotificationSidePanel/index.tsx (hunk 3 above)
Comment (antictor) [reply packages/react-component-library/src/fragments/NotificationSidePanel/index.tsx:20]: Not sure how I would

File: packages/css-framework/src/fragments/_sidebar.scss (hunk 1 above)
Comment (antictor) [reply packages/css-framework/src/fragments/_sidebar.scss]: Helps group the styles to understand the structure

File: packages/react-component-library/src/fragments/NotificationSidePanel/index.tsx (hunk 3 above)
Comment (m7kvqbe1) [reply packages/react-component-library/src/fragments/NotificationSidePanel/index.tsx:20]: You could export `documentClick` and `toggleNotifications` from the hook and just consume like so:

```const [state, setter, somethingElse, somethingElse] = useCustomHook(arg1, arg2, …)```

Maybe it isn’t warranted where there isn’t going to be code re-use?

Just trying to start a discussion around custom hooks usage - I did this in the Pagination implementation in order to encapsulate related logic, even though it was tightly coupled to the component.

Comment (antictor) [reply packages/react-component-library/src/fragments/NotificationSidePanel/index.tsx:20]: Will take a look and do it as a follow up

**Comment by hxltrhuxze** [comment]
Couple of points regarding design.

Can we take a look at the notification sheet? The spacing & colours look a tad off:
Screenshot 2019-07-09 at 12 19 26

Also can we check the spacing between the icons and the text in the sidebar? In the designs, the text label is closer to the icon when the menu is expanded:
Screenshot 2019-07-09 at 12 21 27

**Comment by antictor** [comment]
Updated label spacing and discussed reason for notification popover

**Pull request merged**